blocklylib-core/src/main/java/com/google/blockly/android/codegen/CodeGeneratorManager.java

```diff
@@ -43,54 +53,58 @@ public void onServiceDisconnected(ComponentName arg0) {
         };
     }
 
-    public void onResume() {
-        Intent intent = new Intent(mContext, CodeGeneratorService.class);
-        mContext.bindService(intent, mCodeGenerationConnection, Context.BIND_AUTO_CREATE);
+    /**
+     * Unbind the underlying service (if it is bound).
+     */
+    public void onPause() {
+        mResumed = false;
+        if (isBound()) {
+            mContext.unbindService(mCodeGenerationConnection);
+        }
     }
 
-    public void onPause() {
-        mContext.unbindService(mCodeGenerationConnection);
+    /**
+     * Inform this class that it is ok to bind to the service and remove any stored requests as
+     * the service won't be bound until a new request comes in.
+     */
+    public void onResume() {
+        mResumed = true;
+        mStoredRequests.clear();
     }
 
     /**
      * Calls the Service to request code generation for the workspace passed in.
      *
-     * @param workspace the workspace to request code generation for.
-     * @param codeGenerationCallback the callback to call with the generated code.
+     * @param codeGenerationRequest the request to generate code.
      */
-    public void requestCodeGeneration(List<String> blockDefinitionsJsonPaths,
-                                      List<String> generatorsJsPaths,
-                                      Workspace workspace,
-                                      CodeGeneratorCallback codeGenerationCallback) {
-
-        if (workspace.hasBlocks()) {
-            if (isBound()) {
-                try {
-                    final StringOutputStream serialized = new StringOutputStream();
-                    workspace.serializeToXml(serialized);
-
-                    mGeneratorService.requestCodeGeneration(
-                        new CodeGenerationRequest(
-                            serialized.toString(),
-                            codeGenerationCallback,
-                            blockDefinitionsJsonPaths,
-                            generatorsJsPaths)
-                    );
-                } catch (BlocklySerializerException e) {
-                    Log.wtf(TAG, e);
-                    Toast.makeText(mContext, "Code generation failed.",
-                        Toast.LENGTH_LONG).show();
-
-                }
-            } else {
-                Log.i(TAG, "Generator not bound to service. Skipping run request.");
+    public void requestCodeGeneration(CodeGenerationRequest codeGenerationRequest) {
```

Comment:
Nulls are already handled in the code, but I'll add a null check here so that that's clearer.